tests/integration-tests/tests/cloudwatch_logging/test_cloudwatch_logging.py

```diff
@@ -0,0 +1,460 @@
+
+    def _read_log_configs_from_master(self):
+        """Read the log configs file at /usr/local/etc/cloudwatch_log_files.json."""
+        read_cmd = "cat /usr/local/etc/cloudwatch_log_files.json"
+        config = json.loads(self._run_command_on_master(read_cmd))
+        return config["log_configs"]
+
+    @staticmethod
+    def _clean_log_config(log):
+        """Remove unnecessary fields from the given log dict."""
+        desired_keys = ["file_path", "log_stream_name"]
+        return {key: log[key] for key in desired_keys}
+
+    def _get_relevant_logs(self):
+        """Get subset of all log configs that apply to this cluster's scheduler/os combo."""
+        # Figure out which logs are relevant to the master and computes for this cluster
+        logs = self._read_log_configs_from_master()
+        for log in logs:
+            if self.scheduler not in log["schedulers"] or self.platform not in log["platforms"]:
+                continue
+            for node_role in log["node_roles"]:
+                self._relevant_logs[node_role].append(self._clean_log_config(log))
+        # Give each nodes representative dict in self._cluster_log_state a copy
+        self._cluster_log_state["MasterServer"]["logs"] = {
+            log["file_path"]: log for log in self._relevant_logs["MasterServer"]
+        }
+        for _hostname, compute_instance_dict in self._cluster_log_state["ComputeFleet"].items():
+            compute_instance_dict["logs"] = {
+                log["file_path"]: log.copy() for log in self._relevant_logs["ComputeFleet"]
+            }
+        LOGGER.debug("After populating relevant logs:\n{}".format(self._dump_cluster_log_state()))
+
+    def _run_command_on_master(self, cmd):
+        """Run cmd on cluster's MasterServer."""
+        return self.remote_command_executor.run_remote_command(cmd).stdout.strip()
+
+    def _run_command_on_computes(self, cmd, assert_success=True):
+        """Run cmd on all computes in the cluster."""
+        # Create directory in /shared to direct outputs to
+        out_dir = self._run_command_on_master("mktemp -d -p /shared")
+        redirect = " > {out_dir}/$(hostname -f) ".format(out_dir=out_dir)
+        remote_cmd = cmd.format(redirect=redirect)
+
+        # Run the command, wait for it to complete
+        submit_out = self.scheduler_commands.submit_command(
+            remote_cmd, nodes=self.scheduler_commands.compute_nodes_count()
+        )
+        job_id = self.scheduler_commands.assert_job_submitted(submit_out.stdout)
+        self.scheduler_commands.wait_job_completed(job_id)
+        if assert_success:
+            self.scheduler_commands.assert_job_succeeded(job_id)
+
+        # Read the output and map it to the hostname
+        outputs = {}
+        result_files = self._run_command_on_master("ls {}".format(out_dir))
+        for hostname in result_files.split():
+            path = "/".join((out_dir, hostname))
```

Comment:
Use a path object for `out_dir`, then you can just do `out_dir / hostname`. See https://stackoverflow.com/questions/3751900/create-file-path-from-variables#answer-56620931

Reply:
Done.